core/src/main/java/jenkins/model/OptionalJobProperty.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.model;
+
+import hudson.model.Job;
+import hudson.model.JobProperty;
+import hudson.model.JobPropertyDescriptor;
+import net.sf.json.JSONObject;
+import org.kohsuke.stapler.StaplerRequest;
+
+/**
+ * Job property which may or may not be present.
+ * Must define {@code config-details.jelly} or {@code config-details.groovy}.
+ * May define {@code help.html}.
+ * @since TODO
+ */
+public abstract class OptionalJobProperty<J extends Job<?,?>> extends JobProperty<J> {
+
+    @Override
+    public OptionalJobPropertyDescriptor getDescriptor() {
+        return (OptionalJobPropertyDescriptor) super.getDescriptor();
+    }
+
+    public static abstract class OptionalJobPropertyDescriptor extends JobPropertyDescriptor {
+
+        protected OptionalJobPropertyDescriptor(Class<? extends JobProperty<?>> clazz) {
+            super(clazz);
+        }
+
+        protected OptionalJobPropertyDescriptor() {}
+
+        @Override
+        public JobProperty<?> newInstance(StaplerRequest req, JSONObject formData) throws FormException {
+            return formData.optBoolean("specified") ? super.newInstance(req, formData) : null;
+        }
+
+    }
+
+}
```

Comment:
Yes, be sure to leave line comments on the most upstream applicable PR. I know it is confusing, but not much I can do about it, since I cannot predict if some of the downstream PRs would be delayed or rejected.